File: packages/python-sdk/example/webhook-test-project/usage.py
"""
VaultKey SDK — Example Usage

Run with:
    VAULTKEY_API_KEY=vk_live_... VAULTKEY_API_SECRET=... python examples/usage.py

Make sure VAULTKEY_API_KEY and VAULTKEY_API_SECRET are set in your environment.
"""

import time
import os

from vaultkey import VaultKey, VaultKeyHTTPError

vk = VaultKey(
    api_key=os.environ["VAULTKEY_API_KEY"],
    api_secret=os.environ["VAULTKEY_API_SECRET"],
)


# ---------------------------------------------------------------------------
# Helper: poll a job until completed or failed
# ---------------------------------------------------------------------------

def poll_job(job_id: str, interval: float = 1.0):
    print(f"  Polling job {job_id}...")
    while True:
        result, err = vk.jobs.get(job_id)
        if err:
            raise Exception(f"Job poll failed: {err['message']}")
        print(f"  status: {result['status']}")
        if result["status"] == "completed":
            return result
        if result["status"] == "failed":
            raise Exception(f"Job failed: {result.get('error', 'unknown reason')}")
        time.sleep(interval)


# ---------------------------------------------------------------------------
# 1. Create wallets
# ---------------------------------------------------------------------------

def create_wallets():
    print("\n── Create Wallets ──────────────────────────────────────────")

    evm_wallet, err = vk.wallets.create({
        "user_id": "user_123",
        "chain_type": "evm",
        "label": "Primary EVM wallet",
    })
    if err:
        raise Exception(err["message"])
    print(f"EVM wallet created: {evm_wallet['id']}  {evm_wallet['address']}")

    sol_wallet, err = vk.wallets.create({
        "user_id": "user_123",
        "chain_type": "solana",
    })
    if err:
        raise Exception(err["message"])
    print(f"Solana wallet created: {sol_wallet['id']}  {sol_wallet['address']}")

    return evm_wallet, sol_wallet


# ---------------------------------------------------------------------------
# 2. List wallets for a user
# ---------------------------------------------------------------------------

def list_wallets(user_id: str):
    print("\n── List Wallets ────────────────────────────────────────────")

    result, err = vk.wallets.list_by_user(user_id)
    if err:
        raise Exception(err["message"])

    print(f"Found {len(result['wallets'])} wallet(s) for user {user_id}")
    for w in result["wallets"]:
        print(f"  {w['id']}  {w['chain_type']}  {w['address']}")

    if result["has_more"]:
        print("  (more pages available — use result['next_cursor'] to fetch)")


# ---------------------------------------------------------------------------
# 3. Check balances
# ---------------------------------------------------------------------------

def check_balances(evm_wallet_id: str, sol_wallet_id: str):
    print("\n── Balances ────────────────────────────────────────────────")

    # EVM — prefer chain name over chain ID
    evm_bal, err = vk.wallets.evm_balance(evm_wallet_id, chain_name="base-sepolia")
    if err:
        raise Exception(err["message"])
    print(f"EVM balance: {evm_bal['balance']} {evm_bal['symbol']} on {evm_bal['chain_name']}")

    # Solana
    sol_bal, err = vk.wallets.solana_balance(sol_wallet_id)
    if err:
        raise Exception(err["message"])
    print(f"Solana balance: {sol_bal['balance']} {sol_bal['symbol']}")


# ---------------------------------------------------------------------------
# 4. Sign messages
# ---------------------------------------------------------------------------

def sign_messages(evm_wallet_id: str, sol_wallet_id: str):
    print("\n── Sign Messages ───────────────────────────────────────────")

    # EVM message signing
    evm_job, err = vk.wallets.signing.evm_message(
        evm_wallet_id,
        {
            "payload": {"message": "Hello from VaultKey"},
            "idempotency_key": "sign-evm-001",
        },
    )
    if err:
        raise Exception(err["message"])
    print(f"EVM sign job created: {evm_job['job_id']}")
    evm_result = poll_job(evm_job["job_id"])
    print(f"EVM sign result: {evm_result.get('result')}")

    # Solana message signing
    sol_job, err = vk.wallets.signing.solana_message(
        sol_wallet_id,
        {"payload": {"data": "SGVsbG8gZnJvbSBWYXVsdEtleQ=="}},
    )
    if err:
        raise Exception(err["message"])
    print(f"Solana sign job created: {sol_job['job_id']}")
    poll_job(sol_job["job_id"])


# ---------------------------------------------------------------------------
# 5. Stablecoin transfer
# ---------------------------------------------------------------------------

def stablecoin_transfer(evm_wallet_id: str):
    print("\n── Stablecoin Transfer ─────────────────────────────────────")

    # EVM USDC transfer (gasless — relayer pays gas)
    result, err = vk.stablecoin.transfer(evm_wallet_id, {
        "token": "usdc",
        "to": "0xRecipientAddress",
        "amount": "10.00",
        "chain_type": "evm",
        "chain_name": "base-sepolia",
        "gasless": True,
        "speed": "normal",
        "idempotency_key": "transfer-usdc-001",
    })
    if err:
        raise Exception(err["message"])
    print(f"Transfer job created: {result['job_id']}")
    poll_job(result["job_id"])

    # Check balance after
    bal, _ = vk.stablecoin.balance(evm_wallet_id, {
        "token": "usdc",
        "chain_type": "evm",
        "chain_name": "base-sepolia",
    })
    print(f"USDC balance after transfer: {bal['balance']} {bal['symbol']}")


# ---------------------------------------------------------------------------
# 6. Sweep
# ---------------------------------------------------------------------------

def sweep(evm_wallet_id: str, sol_wallet_id: str):
    print("\n── Sweep ───────────────────────────────────────────────────")

    # EVM sweep
    evm_job, err = vk.wallets.sweep(
        evm_wallet_id,
        {"chain_type": "evm", "chain_name": "base-sepolia"},
    )
    if err:
        raise Exception(err["message"])
    print(f"EVM sweep job: {evm_job['job_id']}")
    poll_job(evm_job["job_id"])

    # Solana sweep
    sol_job, err = vk.wallets.sweep(sol_wallet_id, {"chain_type": "solana"})
    if err:
        raise Exception(err["message"])
    print(f"Solana sweep job: {sol_job['job_id']}")
    poll_job(sol_job["job_id"])


# ---------------------------------------------------------------------------
# 7. List supported chains
# ---------------------------------------------------------------------------

def list_chains():
    print("\n── Supported Chains ────────────────────────────────────────")

    chains, err = vk.chains.list()
    if err:
        raise Exception(err["message"])

    for c in chains:
        label = "(testnet)" if c.get("testnet") else "(mainnet)"
        print(f"  {c['name']:<20} chain_id={c['chain_id']}  {c['native_symbol']}  {label}")


# ---------------------------------------------------------------------------
# Run all examples
# ---------------------------------------------------------------------------

def main():
    print("VaultKey SDK — Example Usage")
    print("=" * 60)

    try:
        evm_wallet, sol_wallet = create_wallets()
        list_wallets("user_123")
        check_balances(evm_wallet["id"], sol_wallet["id"])
        sign_messages(evm_wallet["id"], sol_wallet["id"])
        stablecoin_transfer(evm_wallet["id"])
        sweep(evm_wallet["id"], sol_wallet["id"])
        list_chains()
        print("\n✓ All examples completed.")
    except VaultKeyHTTPError as e:
        print(f"\nAPI error {e.status_code}: {e.error['message']}")
        raise
    except Exception as e:
        print(f"\nError: {e}")
        raise


if __name__ == "__main__":
    main()

File: packages/sdk/examples/usage.ts
/**
 * VaultKey SDK — Example Usage
 *
 * Run with: npx tsx examples/usage.ts
 * Make sure VAULTKEY_API_KEY and VAULTKEY_API_SECRET are set in your environment.
 */

import { VaultKey } from "@vaultkey/sdk";

const vk = new VaultKey({
  apiKey: process.env.VAULTKEY_API_KEY,
  apiSecret: process.env.VAULTKEY_API_SECRET,
});

// ---------------------------------------------------------------------------
// Helper: poll a job until completed or failed
// ---------------------------------------------------------------------------

async function pollJob(jobId: string, intervalMs = 1000) {
  console.log(`  Polling job ${jobId}...`);
  while (true) {
    const { data, error } = await vk.jobs.get(jobId);
    if (error) throw new Error(`Job poll failed: ${error.message}`);
    console.log(`  status: ${data.status}`);
    if (data.status === "completed") return data;
    if (data.status === "failed") throw new Error(`Job failed: ${data.error}`);
    await new Promise((r) => setTimeout(r, intervalMs));
  }
}

// ---------------------------------------------------------------------------
// 1. Create wallets
// ---------------------------------------------------------------------------

async function createWallets() {
  console.log("\n── Create Wallets ──────────────────────────────────────────");

  const { data: evmWallet, error: evmErr } = await vk.wallets.create({
    userId: "user_123",
    chainType: "evm",
    label: "Primary EVM wallet",
  });
  if (evmErr) throw new Error(evmErr.message);
  console.log("EVM wallet created:", evmWallet.id, evmWallet.address);

  const { data: solWallet, error: solErr } = await vk.wallets.create({
    userId: "user_123",
    chainType: "solana",
  });
  if (solErr) throw new Error(solErr.message);
  console.log("Solana wallet created:", solWallet.id, solWallet.address);

  return { evmWallet, solWallet };
}

// ---------------------------------------------------------------------------
// 2. List wallets for a user
// ---------------------------------------------------------------------------

async function listWallets(userId: string) {
  console.log("\n── List Wallets ────────────────────────────────────────────");

  const { data, error } = await vk.wallets.listByUser(userId);
  if (error) throw new Error(error.message);

  console.log(`Found ${data.wallets.length} wallet(s) for user ${userId}`);
  for (const w of data.wallets) {
    console.log(`  ${w.id}  ${w.chainType}  ${w.address}`);
  }

  if (data.hasMore) {
    console.log("  (more pages available — use data.nextCursor to fetch)");
  }
}

// ---------------------------------------------------------------------------
// 3. Check balances
// ---------------------------------------------------------------------------

async function checkBalances(evmWalletId: string, solWalletId: string) {
  console.log("\n── Balances ────────────────────────────────────────────────");

  // EVM — prefer chain name over chain ID
  const { data: evmBal, error: evmErr } = await vk.wallets.evmBalance(
    evmWalletId,
    { chainName: "base-sepolia" }
  );
  if (evmErr) throw new Error(evmErr.message);
  console.log(`EVM balance: ${evmBal.balance} ${evmBal.symbol} on ${evmBal.chainName}`);

  // Solana
  const { data: solBal, error: solErr } = await vk.wallets.solanaBalance(solWalletId);
  if (solErr) throw new Error(solErr.message);
  console.log(`Solana balance: ${solBal.balance} ${solBal.symbol}`);
}

// ---------------------------------------------------------------------------
// 4. Sign messages
// ---------------------------------------------------------------------------

async function signMessages(evmWalletId: string, solWalletId: string) {
  console.log("\n── Sign Messages ───────────────────────────────────────────");

  // EVM message signing
  const { data: evmJob, error: evmErr } = await vk.wallets
    .signing(evmWalletId)
    .evmMessage({
      payload: { message: "Hello from VaultKey" },
      idempotencyKey: "sign-evm-001",
    });
  if (evmErr) throw new Error(evmErr.message);
  console.log("EVM sign job created:", evmJob.jobId);
  const evmResult = await pollJob(evmJob.jobId);
  console.log("EVM sign result:", evmResult.result);

  // Solana message signing
  const { data: solJob, error: solErr } = await vk.wallets
    .signing(solWalletId)
    .solanaMessage({
      payload: { data: Buffer.from("Hello from VaultKey").toString("base64") },
    });
  if (solErr) throw new Error(solErr.message);
  console.log("Solana sign job created:", solJob.jobId);
  await pollJob(solJob.jobId);
}

// ---------------------------------------------------------------------------
// 5. Stablecoin transfer
// ---------------------------------------------------------------------------

async function stablecoinTransfer(evmWalletId: string) {
  console.log("\n── Stablecoin Transfer ─────────────────────────────────────");

  // EVM USDC transfer (gasless — relayer pays gas)
  const { data, error } = await vk.stablecoin.transfer(evmWalletId, {
    token: "usdc",
    to: "0xRecipientAddress",
    amount: "10.00",
    chainType: "evm",
    chainName: "base-sepolia",
    gasless: true,
    speed: "normal",
    idempotencyKey: "transfer-usdc-001",
  });
  if (error) throw new Error(error.message);
  console.log("Transfer job created:", data.jobId);
  await pollJob(data.jobId);

  // Check balance after
  const { data: bal } = await vk.stablecoin.balance(evmWalletId, {
    token: "usdc",
    chainType: "evm",
    chainName: "base-sepolia",
  });
  console.log(`USDC balance after transfer: ${bal.balance} ${bal.symbol}`);
}

// ---------------------------------------------------------------------------
// 6. Sweep
// ---------------------------------------------------------------------------

async function sweep(evmWalletId: string, solWalletId: string) {
  console.log("\n── Sweep ───────────────────────────────────────────────────");

  // EVM sweep
  const { data: evmJob, error: evmErr } = await vk.wallets.sweep(evmWalletId, {
    chainType: "evm",
    chainName: "base-sepolia",
  });
  if (evmErr) throw new Error(evmErr.message);
  console.log("EVM sweep job:", evmJob.jobId);
  await pollJob(evmJob.jobId);

  // Solana sweep
  const { data: solJob, error: solErr } = await vk.wallets.sweep(solWalletId, {
    chainType: "solana",
  });
  if (solErr) throw new Error(solErr.message);
  console.log("Solana sweep job:", solJob.jobId);
  await pollJob(solJob.jobId);
}

// ---------------------------------------------------------------------------
// 7. List supported chains
// ---------------------------------------------------------------------------

async function listChains() {
  console.log("\n── Supported Chains ────────────────────────────────────────");

  const { data: chains, error } = await vk.chains.list();
  if (error) throw new Error(error.message);

  for (const c of chains) {
    const label = c.testnet ? "(testnet)" : "(mainnet)";
    console.log(`  ${c.name.padEnd(20)} chain_id=${c.chainId}  ${c.nativeSymbol}  ${label}`);
  }
}

// ---------------------------------------------------------------------------
// Run all examples
// ---------------------------------------------------------------------------

async function main() {
  console.log("VaultKey SDK — Example Usage");
  console.log("=".repeat(60));

  const { evmWallet, solWallet } = await createWallets();
  await listWallets("user_123");
  await checkBalances(evmWallet.id, solWallet.id);
  await signMessages(evmWallet.id, solWallet.id);
  await stablecoinTransfer(evmWallet.id);
  await sweep(evmWallet.id, solWallet.id);
  await listChains();

  console.log("\n✓ All examples completed.");
}

main().catch((err) => {
  console.error("Error:", err.message);
  process.exit(1);
});